File: services/geminiService.ts
import { GoogleGenAI, GenerateContentResponse, Chat } from "@google/genai";
import { nessieFunctionDeclarations } from "../constants";
import { nessieBankApi } from "./nessieService";

// The API key must be obtained exclusively from the environment variable `process.env.API_KEY`.
const ai = new GoogleGenAI({ apiKey: process.env.API_KEY });

const model = "gemini-2.5-flash";

const systemInstruction = `You are a helpful and professional banking assistant for 'Nessie Bank'.
- Your primary goal is to help users by calling the available functions based on their requests.
- When you need information like an account ID, ask the user for it clearly.
- CRITICAL: User account IDs are numbers, but they may speak them as words (e.g., "one two three four"). You MUST convert these words to digits.
- CRITICAL: The banking system requires account IDs to have the prefix 'acc-'. After converting words to digits, you must prepend 'acc-' to the ID before calling any function. For example, if a user says their ID is "one two three four", you must call the function with accountId: "acc-1234".
- Be concise and clear in your responses.`;


export const geminiService = {
  startChatSession: (): Chat => {
     return ai.chats.create({
      model: model,
      config: {
        systemInstruction: systemInstruction,
        tools: [{ functionDeclarations: nessieFunctionDeclarations }],
      },
    });
  },

  runConversation: async (chat: Chat, userInput: string): Promise<string> => {
    // First, send the user's message
    // FIX: `chat.sendMessage` expects an object with a `message` property. The user input string is passed as the value.
    let response: GenerateContentResponse = await chat.sendMessage({ message: userInput });

    // Check if the model responded with a function call
    if (response.functionCalls && response.functionCalls.length > 0) {
      const functionCall = response.functionCalls[0];
      const functionName = functionCall.name as keyof typeof nessieBankApi;
      
      console.log(`[Gemini] Requested function call: ${functionName}`);
      console.log(`[Gemini] Arguments:`, functionCall.args);

      if (nessieBankApi[functionName]) {
        const args = functionCall.args;
        let functionResult: Record<string, unknown>;

        // FIX: Call API functions with named arguments to avoid issues with argument order from Object.values(). This also resolves the type error.
        switch (functionName) {
          case 'getAccountBalance':
            functionResult = await nessieBankApi.getAccountBalance(args.accountId as string);
            break;
          case 'getRecentTransactions':
            functionResult = await nessieBankApi.getRecentTransactions(args.accountId as string, args.count as number | undefined);
            break;
          case 'transferFunds':
            functionResult = await nessieBankApi.transferFunds(args.fromAccountId as string, args.toAccountId as string, args.amount as number);
            break;
        }
        
        console.log(`[Nessie] Function result:`, functionResult);

        // Send the function's result back to the model
        // FIX: `chat.sendMessage` expects an object with a `message` property. The function response is passed as a `Part` object within an array.
        response = await chat.sendMessage({
          message: [
            {
              functionResponse: {
                name: functionCall.name,
                response: functionResult,
              },
            },
          ],
        });

        // The model will now respond in natural language based on the function result.
        return response.text;
      } else {
        return `I'm sorry, I don't know how to do that. The function ${functionName} is not available.`;
      }
    }

    // If no function call, return the direct text response
    return response.text;
  },
};

File: backend/api/services/gemini_service.py
import os
from typing import Dict, Any, Optional
import google.generativeai as genai
from dotenv import load_dotenv
from .nessie_service import NessieBankService

load_dotenv()

# Configure the Gemini API
genai.configure(api_key=os.getenv('GOOGLE_API_KEY'))

SYSTEM_INSTRUCTION = """You are a helpful and professional banking assistant for 'Nessie Bank'.
- Your primary goal is to help users by calling the available functions based on their requests.
- When you need information like an account ID, ask the user for it clearly.
- CRITICAL: User account IDs are numbers, but they may speak them as words (e.g., "one two three four"). You MUST convert these words to digits.
- CRITICAL: The banking system requires account IDs to have the prefix 'acc-'. After converting words to digits, you must prepend 'acc-' to the ID before calling any function.
- Be concise and clear in your responses."""

class GeminiService:
    def __init__(self):
        self.model = genai.GenerativeModel('gemini-2.5-flash')
        self.nessie_service = NessieBankService()
        self.chat = None
        self.start_chat()

    def start_chat(self):
        """Initialize a new chat session"""
        self.chat = self.model.start_chat(history=[])
        # Send system instruction
        self.chat.send_message(SYSTEM_INSTRUCTION)

    def run_conversation(self, user_input: str, curr_user_name: str, curr_user_id: str, curr_account_id: str) -> str:
        """Process user input and return the response"""
        try:
            # Send user message and get response
            transactions = self.nessie_service.get_recent_transactions(curr_account_id, 10)
            print(f"Recent Transactions for Account {curr_account_id}: {transactions}")
            system_prompt = """
            System prompt: Please analyse the following recent transactions for the user and predict
            to their spending habits, categorizing expenses and identifying any unusual activity. 
            Give them advice on managing their finances based on these transactions. Please respond in 4 sentences.
            """
            
            print("Message sent to Gemini")
            print(system_prompt + " User Input: " + user_input + f" (User: {curr_user_name}, UserID: {curr_user_id}, AccountID: {curr_account_id}), Recent Transactions: {transactions}")
            response = self.chat.send_message(system_prompt + " User Input: " + user_input + f" (User: {curr_user_name}, UserID: {curr_user_id}, AccountID: {curr_account_id}), Recent Transactions: {transactions}")
            

            # Check if we need to call any banking functions
            # This is a simplified version - in practice, you'd need to implement
            # function calling based on the response content
            
            # For now, we'll use basic keyword matching
            response_text = response.text.lower()
            
            
            
            # if 'balance' in response_text:
            #     # Extract account ID from response
            #     # This is a simplified example - you'd need more robust parsing
            #     account_id = self._extract_account_id(response_text)
            #     if account_id:
            #         balance_info = self.nessie_service.get_account_balance(account_id)
            #         return self._format_balance_response(balance_info)
                    
            # elif 'transfer' in response_text:
            #     # Handle transfer logic
            #     # This is a simplified example
            #     from_account, to_account, amount = self._extract_transfer_details(response_text)
            #     if all([from_account, to_account, amount]):
            #         transfer_result = self.nessie_service.transfer_funds(
            #             from_account, to_account, float(amount)
            #         )
            #         return self._format_transfer_response(transfer_result)
            
            return response_text

        except Exception as e:
            print(f"Error in conversation: {str(e)}")
            return "I apologize, but I encountered an error processing your request. Could you please try again?"

    def _extract_account_id(self, text: str) -> Optional[str]:
        """Extract account ID from text - simplified example"""
        # This would need more sophisticated parsing in practice
        import re
        match = re.search(r'acc-(\d+)', text)
        return match.group(0) if match else None

    def _extract_transfer_details(self, text: str) -> tuple[Optional[str], Optional[str], Optional[float]]:
        """Extract transfer details from text - simplified example"""
        # This would need more sophisticated parsing in practice
        # Returns (from_account, to_account, amount)
        return None, None, None

    def _format_balance_response(self, balance_info: Dict[str, Any]) -> str:
        """Format balance information into a response"""
        if balance_info.get('success'):
            return f"Your account balance is ${balance_info['balance']:.2f}"
        return "I'm sorry, I couldn't retrieve your balance at this time."

    def _format_transfer_response(self, transfer_result: Dict[str, Any]) -> str:
        """Format transfer result into a response"""
        if transfer_result.get('success'):
            return "The transfer was completed successfully."
        return f"I'm sorry, the transfer couldn't be completed: {transfer_result.get('error', 'Unknown error')}"

gemini_service = GeminiService()
